source-code/core/src/lint/linter.ts

```diff
@@ -1,235 +1,166 @@
-
-const getResourceForLanguage = (resources: Resource[], language: string) =>
+import type { Config } from "@inlang/core/config"
+import type * as ast from "@inlang/core/ast"
+import type { LintedResource, LintRule, Visitors } from "./rule.js"
+import { createReportFunction } from "./report.js"
+
+const getResourceForLanguage = (resources: ast.Resource[], language: string) =>
 	resources.find(({ languageTag }) => languageTag.name === language)
 
+/**
+ * Lints the given resources.
+ *
+ * The linted resources will be returned with lint information added to each node.
+ * The returned errors are unexpected errors that occurred during linting, not
+ * lint errors themselves.
+ *
+ * @example
+ *   const [lintedResources, errors] = await lint({ config, resources })
+ *   if (errors) {
+ *     // handle unexpected errors during the linting process.
+ *     // this errors are not lint errors themselves!
+ *   }
+ *   const lints = getLintReports(lintedResources, { options })
+ */
 export const lint = async (args: {
 	config: Pick<Config, "lint" | "languages" | "referenceLanguage">
-	resources: Resource[]
-}): Promise<LintedResource[]> => {
+	resources: ast.Resource[]
+}): Promise<[lintedResources: LintedResource[], errors?: Error[]]> => {
 	const { referenceLanguage, languages, lint } = args.config
-	// linting the resources should not modify args.resources.
 	const resources = structuredClone(args.resources)
 	if (lint === undefined || lint.rules.length === 0) {
-		return args.resources
+		return [args.resources, []]
 	}
 	const reference = getResourceForLanguage(resources, referenceLanguage)
+	const errors: Error[] = []
 
 	await Promise.all(
-		lint.rules.flat().map((lintRule) =>
+		lint.rules.flat().map((rule) =>
 			processLintRule({
-				lintRule,
+				rule,
 				referenceLanguage,
 				languages,
 				reference,
 				resources,
-			}).catch((e) => console.error(`Unexpected error in lint rule '${lintRule.id}':`, e)),
+			}).catch((e) =>
+				errors.push(new Error(`Unexpected error in lint rule '${rule.id}'`, { cause: e })),
+			),
 		),
 	)
 
-	return resources as LintedResource[]
+	return [resources as LintedResource[], errors.length > 0 ? errors : undefined]
 }
 
-const processLintRule = async ({
-	lintRule,
-	referenceLanguage,
-	languages,
-	reference,
-	resources,
-}: {
-	lintRule: LintRule
+const processLintRule = async (args: {
+	rule: LintRule
 	referenceLanguage: string
 	languages: string[]
-	reference: Resource | undefined
-	resources: Resource[]
+	reference: ast.Resource | undefined
+	resources: ast.Resource[]
 }) => {
-	const { level, id, setup, visitors, teardown } = lintRule
-	if (!level) return
-
-	const context = createContext(id, level)
+	const { referenceLanguage, languages, reference, resources } = args
 
-	const payload = await setup({ referenceLanguage, languages, context })
+	const report = createReportFunction(args.rule)
+	const { visitors } = await args.rule.setup({ config: { referenceLanguage, languages }, report })
```

Comment:
I would just pass the props needed for the `createReportFunction`
```suggestion
	const  { id, level, setup } = args.rule
	const report = createReportFunction({ id, level })
	const { visitors } = await setup({ config: { referenceLanguage, languages }, report })
```

Reply:
I stopped destructering props because SolidJS looses reactivety on desctructering (proxy are not desctructed :/)